### PR TITLE
IP Range and CIDR support

### PR DESCRIPTION
**Pull request by divfor**
An alternative PR for using IP range option #1293 , no Network Interface required.
Example of run script:
```bash
#!/bin/bash
s=${1:-"ax.js:1000:10m:rps_http"}
read jsfile vus duration casename <<<${s//:/ }
shift
echo -e "testfile[$jsfile] testcase[$casename] vus[$vus] duration[$duration]\noptions[$*]"

set_k6_local_route(){ #$1:net $2:ifname
  net="$1" ifname=${2-"lo"}
  ip r s t local |grep -q "^local $net" || \
  ip -6 r s t local |grep -q "^local $net" || \
  ip route add local "$net" dev $ifname
  ip r s t local |grep "^local $net" || ip -6 r s t local |grep "^local $net"
}

# k6 client ip -- use local type route instead
set_k6_local_route "10.50.0.0/16" "eth0"
set_k6_local_route "fd00:0:1::0/120" "eth0"

# Linux: echo "ulimit -n 1000000" >> ~/.bashrc
# MacOS: sudo launchctl limit maxfiles 65536 1000000
ulimit -n 1000000

# run k6 for ipv4 targets
TESTCASE=$casename VUS=$vus DURATION=$duration k6 run --ip="10.50.0.62-10.50.2.100" --verbose --no-thresholds $* $jsfile
# run k6 for ipv6 targets
sysctl -w net.ipv6.ip_nonlocal_bind=1
TESTCASE=$casename VUS=$vus DURATION=$duration k6 run --ip="fd00:0:188::/120" --verbose --no-thresholds $* $jsfile
sysctl -w net.ipv6.ip_nonlocal_bind=0
```

### CODE REVIEW

File: js/runner.go
Comment (divfor) [inline js/runner.go:259]: GetIpPool() should be called once in SetOptions, but I am not familiar with that code area to support it on both CLI and Config file...

File: js/runner.go
Comment (MStoykov) [inline js/runner.go:205]: I don't think I like that :(. 

You might have seen this in the wild and maybe this is "common" practice but given how ipv6 is supposed to be structured I find it very unlikely that you will get a mask that is ... THIS big :D 

And as I've argued before we probably should give IPs to VUs in some orderly way so that if you have 20 IPs and 40 VUs -  each IP will be used by 2 VUs which currently doesn't seem like something that will happen.

**Review by MStoykov — CHANGES_REQUESTED** [review]
LGTM so far!
There are a couple of things that I think must be changed before we can merge this :
1. This shouldn't be configurable from the script or a config file as I don't think this will be portable or all the useful. 
2. Support a single IP. 

And a couple that I would like, but arguably can be changed later and it can be done in a separate PR.
1. support multiple ranges at the same time 
2. getting the IP should not be randomized between the VUs IMO, it should be ordered as that will give better and more easily understandable scalability.

I would recommend you wait for a day or two and rebase on master when we have merged the Github Actiosn PR as well.  

Maybe some more comments around the IP range calculations will be nice too.

Great job and thansk again for this PR :tada:

File: js/runner.go
Comment (divfor) [reply js/runner.go:198]: 1. sometimes VUs number is more than IP addrs available so they must share IPs, that is why I use rand() % ip_num.   for 1:1 ordered binding, specify a mode to not using rand() but only using vu_id % ip_num?
2. instead of calculating, unrolling all range IPs to a linear array may require too much memory if the range is very big - especially with CIDR format. 
3. for multiple ranges, how to map VUs to all possible IPs? the later ranges will be less possibility to be picked if append to array list.

File: js/runner.go
Comment (divfor) [reply js/runner.go:216]: this is used to test IPv6 only - both of network part and host part should has a range to change. typically, for load balance algorithm like RSS that requires a good hash key including both of network and host bytes. if you don't need this feature, just set a mask >= 64 or a small range.

Comment (divfor) [reply js/runner.go:216]: from my case, I often use 40K IPs with one Linux machine:)

File: js/runner.go
Comment (divfor) [reply js/runner.go:137]: yes, if use one IP to specify, this code requires --ip=10.10.10.1-10.10.10.1, or remove this option and K6 will pick up interface's primary IP

File: lib/options.go
Comment (divfor) [reply lib/options.go:387]: it is ok for me  -- I copy this but not quite sure what it will do.

File: js/runner.go
Comment (divfor) [reply js/runner.go:198]: @MStoykov Will VUs be able to change their local IP after initialisation? looks like NewVU() is called only in scale up stage

File: js/runner.go
Comment (divfor) [reply js/runner.go:216]: // IPv4: one IP block from start IP to end IP
// IPv6: two independent IP blocks: high 64-bit for net and low 64-bit for hosts
// This is useful if you want to travel some IPv6 nets without traveling 2^64 hosts

Comment (MStoykov) [reply js/runner.go:216]: As I explained above ... having 2^65 ips for *outgoing* requests ... seems impossible at least for anything that k6 is currently runnig on. Even at 40k IPs ... your OS is probably using more memory just to configure this that k6 will use to unroll them :). 

I might be missing some strange use case and way to do this but as far as I can see ... while you can having full subnet for incoming requests with [anyip](https://blog.widodh.nl/2016/04/anyip-bind-a-whole-subnet-to-your-linux-machine/) this seems to not work for outgoing ones ... which more or less removes the need for k6 to support subnets of the size of /65+, actually a lot less as I would expect the resource usage of a OS to overwhelm thsi pretty fast.

Also this is pretty impractical for k6 ... as you are unlikely to have enough VUs to use even a fraction of the IPs  on a single machine (I would really like it if it was possible to run 2^20 VUs .. but even with 1mb of memory per each this is like 1TB of memory ... just for the VUs which still means that the unrolling of the IPs will be a drop in the ocean of memory used

Comment (divfor) [reply js/runner.go]: local type route is also usable for k6 outgoing requests, so no need to setup IPs to linux kernel -- actually I was using it for a while using serval GB memory.

Comment (divfor) [reply js/runner.go]: here is my K6 using 4k VUs:
image
image

File: js/runner.go
Comment (divfor) [reply js/runner.go]: it can match requirements like:
1. multiple IPs: ip1,ip2,ip3
2. multiple IPs and a range: ip1,ip2,ip3,ip4-ip10|0|6
3. multiple ranges: ip1-ip8,ip10/29

File: cmd/options.go
Comment (divfor) [reply cmd/options.go:94]: now they are all supported

File: js/runner.go
Comment (MStoykov) [reply js/runner.go]: Can you elaborate on how you actually managed to get requests out of a /65 ipv6 network ... for example

**Comment by divfor** [comment]
maybe I did not talk it clear: 

The weight of each IPBlock is actually the number IPs planned to use:
VU id will loop back to the first IPBlock when total planed IPs number (it is all weights' sum) are picked up, so by default it is sequential for IPBlocks (also for single IPs). 

sequential and random are only subject to inside IPBlock:
when picked an IPBlock, mode will decide which IP will be selected: by default it is LoopIncSelectIP (VU id % range-num), or set to RandomSelectIP for random
Examples could verify from last commit:

1. all single sequential:  1.1.1.1,2.2.2.2,3.3.3.3
2. sequential 3 single IPs and then sequential 9 range IPs: 1.1.1.1,2.2.2.2,3.3.3.1-3.3.3.9
3. sequential 3 single IPs and then sequential 256 cidr IPs: 1.1.1.1,2.2.2.2,3.3.3.0/24
4. sequential 3 single IPs and then sequential 10 cidr IPs: 1.1.1.1,2.2.2.2,3.3.3.0/24|0|10
5. sequential 3 single IPs and then random 10 cidr IPs: 1.1.1.1,2.2.2.2,3.3.3.0/24|1|10

btw, IPv6 has similar usage

File: js/runner.go
Comment (divfor) [reply js/runner.go]: For example, a /65 IPv6 range format like fd00:1:2:0::0-fd00:1:2:ff::3ff means:
network will be range of 0 ~ 255 and selected by VU_id % IPBlock.netN
host will be range of 0 ~ 1023 and selected by VU_id % IPBlock.hostN

Of cause, if one don't want so strange range method(it is most like case to test load balancer), just give host part for range or give a CIDR mask <= /64

Comment (MStoykov) [reply js/runner.go]: I just now noticed that you have actually explained how to make the setup in the PR description :facepalm: 
I did manage to add /31 network (I only have ipv4 and didn't want to break my network configuring stuff to test this :D ) and it seems to work ... I am still somewhat skeptic that it will be a common practice to use that on load generator machine ... but it's fine to support it

File: js/iprange.go
Comment (MStoykov) [inline js/iprange.go]: This will need to have better user-facing names, even if they are only two see [compatibilityMode](https://github.com/loadimpact/k6/blob/master/lib/runtime_options.go#L30-L40) for how it's done in other places

File: js/iprange_test.go
Comment (MStoykov) [inline js/iprange_test.go:85]: But ... there are 254 actuallu usable addresses ... and the code doesn't return `.0` and `.255` unless they are the only IP ... which I think is the correct behaviour. But than this 256 is wrong

File: js/iprange.go
Comment (MStoykov) [inline js/iprange.go:1]: This whole file should be in lib/types

**Review by MStoykov — COMMENTED** [review]
The more I look at the code ... the more it looks like ... you've added the options but didn't implement them 
- weights is always one
- the mode is always loopIncSelect

and the more it seems that we wouldn't have time for most of those.

So at this point in time I am of the opinion that it will be better to scale this down to just single ips and get that inside (no weights, no random) and try to work on ranges and cidr networks after that(which will likely push them off until v0.30.0)

The reason for dropping ranges and (especially) cidrs is that looking at the math required ... I do think this will take a very long time to review and I much more prefer to have at least some multi IP support that we can expand on ... instead of nothing.

But even if you are very determent to have most of the other things ... I don't think weights should be part of this release at all (especially not in the way that you described them) and the mode needs to be between all the IPs ... 

Given the fact that we have 2 weeks to get this merged (I don't want this to be in the last week before the release) and the need for testing of this (which will probably take longer than even I expect) I am going to have to take over if you don't scale this back down, sorry. There just isn't enough time for me to look into everything around how the math (should) work and then try to figure out how to document it on top of all the other things in this PR and the others I have to work on.

I hope you agree to scale this down and work on it, but regardless you will be credited and will be welcome to add whatever is dropped from this PR in the future :)

File: js/iprange.go
Comment (divfor) [reply js/iprange.go]: any suggestion for the names?

File: js/runner.go
Comment (divfor) [reply js/runner.go]: now I move the parsing into newRunner()

File: js/iprange_test.go
Comment (divfor) [reply js/iprange_test.go:85]: yes. this just tests for board checking. in practice a user should use  "ipv4 cidr 1":  {254, 1, "192.168.0.01/24|0|254"}

File: js/iprange.go
Comment (divfor) [reply js/iprange.go]: done

**Comment by divfor** [comment]
Actually I have tested and confirmed all have been featured with mixed of range/weight/random.
Okey, let's focus to simply the release notes:  only support single IPs and range/cidr, no weights no random

which one do you prefer to ?
1. keep current commit code, only verify the supported option format
2. I remove current commit code and re-write a simple version and test them, re-submit a PR

**Comment by MStoykov** [comment]
I think it will be better if there is a new PR with ... less commits :)
I recommend that you
1. rebase on the latest master 
2. configure your ide/text editor for golang development as that will automatically run for you gofmt/goimports.
3. run golangci-lint as in ` golangci-lint run --new-from-rev=origin/master` I have this as a pre-push hook in git:
```
#!/bin/sh

# An example hook script to verify what is about to be pushed.  Called by "git
# push" after it has checked the remote status, but before anything has been
# pushed.  If this script exits with a non-zero status nothing will be pushed.
#
# This hook is called with the following parameters:
#
# $1 -- Name of the remote to which the push is being done
# $2 -- URL to which the push is being done
#
# If pushing without using a named remote those arguments will be equal.
#
# Information about the commits which are being pushed is supplied as lines to
# the standard input in the form:
#
#   <local ref> <local sha1> <remote ref> <remote sha1>
#
# This sample shows how to prevent push of commits where the log message starts
# with "WIP" (work in progress).

remote="$1"
url="$2"

z40=0000000000000000000000000000000000000000

while read local_ref local_sha remote_ref remote_sha
do
        if [ "$local_sha" = $z40 ]
        then
                # Handle delete
                :
        else
                if [ "$remote_sha" = $z40 ]
                then
                        # New branch, examine all commits
                        since="origin/master"
                else
                        # Update to existing branch, examine new commits
                        since="$remote_sha"
                fi

        exec golangci-lint run --new-from-rev=$since
        fi
done

exit 0
```

**Comment by divfor** [comment]
okey, let me try to follow up this -- as I am a 3 weeks golang newbie from a QA tester using k6...

**Comment by divfor** [comment]
closed and move #1674 for less commit

**Pull request closed without merging**
